lib/database.cpp

```diff
@@ -0,0 +1,240 @@
+// SPDX-FileCopyrightText: 2021 Tobias Fella <fella@posteo.de>
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "database.h"
+
+#include <QtSql/QSqlDatabase>
+#include <QtSql/QSqlQuery>
+#include <QtSql/QSqlError>
+#include <QtCore/QStandardPaths>
+#include <QtCore/QDebug>
+#include <QtCore/QDir>
+
+#include "e2ee/e2ee.h"
+#include "e2ee/qolmsession.h"
+#include "e2ee/qolminboundsession.h"
+
+//TODO: delete room specific data when leaving room
+
+using namespace Quotient;
+Database::Database()
+{
+    QSqlDatabase::addDatabase(QStringLiteral("QSQLITE"));
+    QString databasePath = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
+    QDir(databasePath).mkpath(databasePath);
+    QSqlDatabase::database().setDatabaseName(databasePath + QStringLiteral("/database.db3"));
+    QSqlDatabase::database().open();
+
+    switch(version()) {
+        case 0: migrateTo1();
+    }
+}
+
+int Database::version()
+{
+    auto query = execute(QStringLiteral("PRAGMA user_version;"));
+    if (query.next()) {
+        bool ok;
+        int value = query.value(0).toInt(&ok);
+        qDebug() << "Database version" << value;
+        if (ok)
+            return value;
+    } else {
+        qCritical() << "Failed to check database version";
+    }
+    return -1;
+}
+
+QSqlQuery Database::execute(const QString &queryString)
+{
+    auto query = QSqlDatabase::database().exec(queryString);
+    if (query.lastError().type() != QSqlError::NoError) {
+        qCritical() << "Failed to execute query";
+        qCritical() << query.lastQuery();
+        qCritical() << query.lastError();
+    }
+    return query;
+}
+
+QSqlQuery Database::execute(QSqlQuery &query)
+{
+    if (!query.exec()) {
+        qCritical() << "Failed to execute query";
+        qCritical() << query.lastQuery();
+        qCritical() << query.lastError();
+    }
+    return query;
+}
+
+void Database::transaction()
+{
+    QSqlDatabase::database().transaction();
+}
+
+void Database::commit()
+{
+    QSqlDatabase::database().commit();
+}
+
+void Database::migrateTo1()
+{
+    qDebug() << "Migrating database to version 1";
+    transaction();
+    execute(QStringLiteral("CREATE TABLE Accounts (matrixId TEXT UNIQUE, pickle TEXT);"));
```

Comment:
Wouldn't it be a good idea to add some indexes? This should improve the performance a bit and also make sure that some values are unique

Reply:
also by conventions in SQL the table names and row names are usually in snake_case. But this a nitpick

Reply:
table names are changed, i would put indexes in the backlog